### PR TITLE
prefer class methods over instance methods and fix the addon availability check

### PR DESCRIPTION
**Pull request by MaximilianMeister**
needs (partial) backport

### CODE REVIEW

File: crowbar_framework/app/models/api/node.rb [hunk 1]
```diff
@@ -28,7 +28,7 @@ def repocheck(options = {})
             "available" => true,
             "repos" => {}
           }
-          platform = Api::Upgrade.new.target_platform(platform_exception: addon)
+          platform = Api::Upgrade.target_platform(platform_exception: addon)
```
Comment (jsuchome) [inline crowbar_framework/app/models/api/node.rb]: This would ideally belong to previous commit

File: crowbar_framework/spec/models/api/upgrade_spec.rb [hunk 2]
```diff
@@ -89,7 +85,14 @@
         receive(:find).with("state:crowbar_upgrade AND NOT run_list_map:pacemaker-cluster-member").
         and_return([])
       )
-      expect(subject.services).to eq([:ok, ""])
+      allow_any_instance_of(NodeObject).to receive(:ssh_cmd).with(
+        "/usr/sbin/crowbar-shutdown-services-before-upgrade.sh"
+      ).and_return(true)
+      allow_any_instance_of(NodeObject).to receive(:ssh_cmd).with(
+        "/usr/sbin/crowbar-shutdown-services-before-upgrade.sh"
+      ).and_return(true)
```
Comment (jsuchome) [inline crowbar_framework/spec/models/api/upgrade_spec.rb]: It would probably be better to differentiate between these 2 calls: both is executed on different set of nodes.

**Comment by MaximilianMeister** [comment]
<a href="https:&#x2F;&#x2F;trello.com&#x2F;c&#x2F;8X7elkka&#x2F;82-cleanup-upgrade-related-mess"> Cleanup upgrade related mess</a>

File: crowbar_framework/spec/models/api/upgrade_spec.rb (hunk 2 above)
Comment (MaximilianMeister) [inline crowbar_framework/spec/models/api/upgrade_spec.rb]: done in the latest commit

Comment (jsuchome) [inline crowbar_framework/spec/models/api/upgrade_spec.rb]: thanks!

Comment (aspiers) [inline crowbar_framework/spec/models/api/upgrade_spec.rb]: This is the kind of thing which ideally would belong in a separate PR - IIUC it's independent of the switch to class methods.

File: crowbar_framework/app/models/api/crowbar.rb [hunk 3]
```diff
@@ -18,11 +18,11 @@
 
 module Api
   class Crowbar < Tableless
-    attr_reader :version
+    attr_reader :version, :addons
 
     def initialize
       @version = ENV["CROWBAR_VERSION"]
-      @addons = addons
+      @addons = features
```
Comment (aspiers) [inline crowbar_framework/app/models/api/crowbar.rb]: Why do this in `#initialize`?  If you're really worried about the calculation being repeated multiple times in a single request, then simply memoize the result the first time it's calculated, so that it's only calculated when it's really needed.

File: crowbar_framework/app/models/api/upgrade.rb [hunk 4]
```diff
@@ -48,7 +48,7 @@ def repocheck
       def target_platform(options = {})
         platform_exception = options.fetch(:platform_exception, nil)
 
-        case crowbar.version
+        case ENV["CROWBAR_VERSION"]
```
Comment (aspiers) [inline crowbar_framework/app/models/api/upgrade.rb]: Presumably this also eliminates the need to cache the value in `#initialize`?

**Comment by aspiers** [comment]
BTW by "this removes DRY" I presume you mean "this obeys DRY and removes duplication"?  We don't want to remove the DRY rule, ever! ;-)

**Comment by MaximilianMeister** [comment]
> So I would kindly ask, please keep commits as small as possible, and PRs as small as possible.

will do, I agree. If I would have had to review it, I would have run into the same issues you mentioned I guess
